File: services/lineage-service/src/iceberg_schema.rs
//! Iceberg schema substrate for `of_lineage.*` (S5.2.a).
//!
//! Pins the column set, partition transform and sort order for the
//! three tables behind the OpenLineage materialisation pipeline. The
//! writer is in [`crate::kafka_to_iceberg`] (lands in the runtime PR).
//!
//! Table layout follows OpenLineage 1.x event conventions
//! (<https://openlineage.io/spec/>) trimmed to what the Foundry-parity
//! UI actually queries.

pub mod runs {
    use super::common;

    pub const TABLE: &str = "runs";

    pub mod fields {
        pub const RUN_ID: &str = "run_id"; // uuid, OpenLineage run id
        pub const JOB_NAMESPACE: &str = "job_namespace";
        pub const JOB_NAME: &str = "job_name";
        pub const STARTED_AT: &str = "started_at"; // partition source
        pub const COMPLETED_AT: &str = "completed_at";
        pub const STATE: &str = "state"; // RUNNING|COMPLETE|FAIL|ABORT
        pub const FACETS: &str = "facets"; // serialized JSON
    }

    pub const PARTITION_SOURCE_FIELD: &str = fields::STARTED_AT;
    pub const PARTITION_TRANSFORM: &str = common::DAY;
    pub const SORT_FIELD: &str = fields::STARTED_AT;
    pub const SORT_DIRECTION: &str = common::ASC;
    pub const REQUIRED: &[&str] = &[
        fields::RUN_ID,
        fields::JOB_NAMESPACE,
        fields::JOB_NAME,
        fields::STARTED_AT,
        fields::STATE,
    ];
}

pub mod events {
    use super::common;

    pub const TABLE: &str = "events";

    pub mod fields {
        pub const EVENT_ID: &str = "event_id"; // uuid, dedup
        pub const RUN_ID: &str = "run_id"; // uuid, FK→runs.run_id
        pub const EVENT_TIME: &str = "event_time"; // partition source
        pub const EVENT_TYPE: &str = "event_type"; // START|RUNNING|COMPLETE|FAIL|ABORT
        pub const PRODUCER: &str = "producer"; // OL producer URI
        pub const SCHEMA_URL: &str = "schema_url";
        pub const PAYLOAD: &str = "payload"; // serialized JSON
    }

    pub const PARTITION_SOURCE_FIELD: &str = fields::EVENT_TIME;
    pub const PARTITION_TRANSFORM: &str = common::DAY;
    pub const SORT_FIELD: &str = fields::EVENT_TIME;
    pub const SORT_DIRECTION: &str = common::ASC;
    pub const REQUIRED: &[&str] = &[
        fields::EVENT_ID,
        fields::RUN_ID,
        fields::EVENT_TIME,
        fields::EVENT_TYPE,
    ];
}

pub mod datasets_io {
    use super::common;

    pub const TABLE: &str = "datasets_io";

    pub mod fields {
        pub const RUN_ID: &str = "run_id";
        pub const EVENT_TIME: &str = "event_time"; // partition source
        pub const SIDE: &str = "side"; // "input" | "output"
        pub const DATASET_NAMESPACE: &str = "dataset_namespace";
        pub const DATASET_NAME: &str = "dataset_name";
        pub const FACETS: &str = "facets"; // JSON
    }

    pub const PARTITION_SOURCE_FIELD: &str = fields::EVENT_TIME;
    pub const PARTITION_TRANSFORM: &str = common::DAY;
    pub const SORT_FIELD: &str = fields::EVENT_TIME;
    pub const SORT_DIRECTION: &str = common::ASC;
    pub const REQUIRED: &[&str] = &[
        fields::RUN_ID,
        fields::EVENT_TIME,
        fields::SIDE,
        fields::DATASET_NAMESPACE,
        fields::DATASET_NAME,
    ];
    pub const SIDE_INPUT: &str = "input";
    pub const SIDE_OUTPUT: &str = "output";
}

pub mod common {
    pub const DAY: &str = "day";
    pub const ASC: &str = "asc";
    pub const NULLS_LAST: &str = "nulls-last";

    /// Lineage tables are smaller than audit and we *do* allow snapshot
    /// expiration after 90 days — lineage history is regenerable from
    /// Cassandra hot path, unlike audit. This contrasts with
    /// `audit_sink::iceberg_schema::retention` which forbids expiry.
    pub const TABLE_PROPERTIES: &[(&str, &str)] = &[
        ("write.format.default", "parquet"),
        ("write.parquet.compression-codec", "zstd"),
        ("history.expire.max-snapshot-age-ms", "7776000000"), // 90d
        ("history.expire.min-snapshots-to-keep", "10"),
    ];
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::kafka_to_iceberg::iceberg_target;

    #[test]
    fn table_names_match_target_constants() {
        assert_eq!(runs::TABLE, iceberg_target::TABLE_RUNS);
        assert_eq!(events::TABLE, iceberg_target::TABLE_EVENTS);
        assert_eq!(datasets_io::TABLE, iceberg_target::TABLE_DATASETS_IO);
    }

    #[test]
    fn partition_transform_consistent_with_target() {
        // events.event_time is the canonical partition source quoted in
        // the target const `"day(event_time)"`.
        assert_eq!(
            format!("{}({})", events::PARTITION_TRANSFORM, events::PARTITION_SOURCE_FIELD),
            iceberg_target::PARTITION_TRANSFORM
        );
    }

    #[test]
    fn datasets_io_side_values_documented() {
        assert_eq!(datasets_io::SIDE_INPUT, "input");
        assert_eq!(datasets_io::SIDE_OUTPUT, "output");
    }
}

File: openfoundry-go/services/lineage-service/internal/icebergschema/icebergschema.go
// Package icebergschema pins the column set, partition transform
// and sort order for the three Iceberg tables behind the
// OpenLineage materialisation pipeline. The writer ports in a
// follow-up slice; pinning these constants now keeps the wire-
// compat tests honest.
//
// Table layout follows OpenLineage 1.x event conventions
// (https://openlineage.io/spec/) trimmed to what the Foundry-parity
// UI actually queries.
package icebergschema

// --- Common ---------------------------------------------------------------

const (
	Day       = "day"
	Asc       = "asc"
	NullsLast = "nulls-last"
)

// TableProperties are applied to every materialised lineage table.
// Lineage history is regenerable from the Cassandra hot path, so
// 90-day snapshot expiry is allowed (contrast with audit_sink which
// forbids expiry entirely).
var TableProperties = [][2]string{
	{"write.format.default", "parquet"},
	{"write.parquet.compression-codec", "zstd"},
	{"history.expire.max-snapshot-age-ms", "7776000000"}, // 90d
	{"history.expire.min-snapshots-to-keep", "10"},
}

// --- Runs table -----------------------------------------------------------

const RunsTable = "runs"

// Field names. Pinned because downstream readers (UI, lineage-graph
// query layer) reference them by literal string.
const (
	RunsFieldRunID        = "run_id"
	RunsFieldJobNamespace = "job_namespace"
	RunsFieldJobName      = "job_name"
	RunsFieldStartedAt    = "started_at"
	RunsFieldCompletedAt  = "completed_at"
	RunsFieldState        = "state"
	RunsFieldFacets       = "facets"
)

// Field IDs. Iceberg field IDs are part of the on-disk format —
// changing one is a metadata-rewrite operation.
const (
	RunsFieldIDRunID        int32 = 1
	RunsFieldIDJobNamespace int32 = 2
	RunsFieldIDJobName      int32 = 3
	RunsFieldIDStartedAt    int32 = 4
	RunsFieldIDCompletedAt  int32 = 5
	RunsFieldIDState        int32 = 6
	RunsFieldIDFacets       int32 = 7
)

// Partition + sort.
const (
	RunsPartitionSourceField = RunsFieldStartedAt
	RunsPartitionTransform   = Day
	RunsSortField            = RunsFieldStartedAt
	RunsSortDirection        = Asc
)

// RunsRequired pins the NOT NULL columns.
var RunsRequired = []string{
	RunsFieldRunID,
	RunsFieldJobNamespace,
	RunsFieldJobName,
	RunsFieldStartedAt,
	RunsFieldState,
}

// --- Events table ---------------------------------------------------------

const EventsTable = "events"

const (
	EventsFieldEventID   = "event_id"
	EventsFieldRunID     = "run_id"
	EventsFieldEventTime = "event_time"
	EventsFieldEventType = "event_type"
	EventsFieldProducer  = "producer"
	EventsFieldSchemaURL = "schema_url"
	EventsFieldPayload   = "payload"
)

const (
	EventsFieldIDEventID   int32 = 1
	EventsFieldIDRunID     int32 = 2
	EventsFieldIDEventTime int32 = 3
	EventsFieldIDEventType int32 = 4
	EventsFieldIDProducer  int32 = 5
	EventsFieldIDSchemaURL int32 = 6
	EventsFieldIDPayload   int32 = 7
)

const (
	EventsPartitionSourceField = EventsFieldEventTime
	EventsPartitionTransform   = Day
	EventsSortField            = EventsFieldEventTime
	EventsSortDirection        = Asc
)

var EventsRequired = []string{
	EventsFieldEventID,
	EventsFieldRunID,
	EventsFieldEventTime,
	EventsFieldEventType,
}

// --- Datasets I/O table ---------------------------------------------------

const DatasetsIOTable = "datasets_io"

const (
	DatasetsIOFieldRunID            = "run_id"
	DatasetsIOFieldEventTime        = "event_time"
	DatasetsIOFieldSide             = "side"
	DatasetsIOFieldDatasetNamespace = "dataset_namespace"
	DatasetsIOFieldDatasetName      = "dataset_name"
	DatasetsIOFieldFacets           = "facets"

	DatasetsIOSideInput  = "input"
	DatasetsIOSideOutput = "output"
)

const (
	DatasetsIOFieldIDRunID            int32 = 1
	DatasetsIOFieldIDEventTime        int32 = 2
	DatasetsIOFieldIDSide             int32 = 3
	DatasetsIOFieldIDDatasetNamespace int32 = 4
	DatasetsIOFieldIDDatasetName      int32 = 5
	DatasetsIOFieldIDFacets           int32 = 6
)

const (
	DatasetsIOPartitionSourceField = DatasetsIOFieldEventTime
	DatasetsIOPartitionTransform   = Day
	DatasetsIOSortField            = DatasetsIOFieldEventTime
	DatasetsIOSortDirection        = Asc
)

var DatasetsIORequired = []string{
	DatasetsIOFieldRunID,
	DatasetsIOFieldEventTime,
	DatasetsIOFieldSide,
	DatasetsIOFieldDatasetNamespace,
	DatasetsIOFieldDatasetName,
}
